packages/core/src/modules/mdoc/MdocContext.ts

```diff
@@ -0,0 +1,136 @@
+import type { AgentContext } from '../../agent'
+import type { JwkJson } from '../../crypto'
+import type { MdocContext, X509Context } from '@protokoll/mdoc-client'
+
+import { p256 } from '@noble/curves/p256'
+import { hkdf } from '@noble/hashes/hkdf'
+import { sha256 } from '@noble/hashes/sha2'
+import * as x509 from '@peculiar/x509'
+import { exportJwk, importX509, verifyWithJwk } from '@protokoll/crypto'
+
+import { CredoWebCrypto, getJwkFromJson, Hasher } from '../../crypto'
+import { Buffer, TypedArrayEncoder } from '../../utils'
+import { X509Certificate, X509Service } from '../x509'
+
+export const getMdocContext = (agentContext: AgentContext): MdocContext => {
+  const crypto = new CredoWebCrypto(agentContext)
+  return {
+    crypto: {
+      digest: async (input) => {
+        const { bytes, digestAlgorithm } = input
+        return new Uint8Array(crypto.digest(digestAlgorithm, bytes))
+      },
+      random: (length) => {
+        return crypto.getRandomValues(new Uint8Array(length))
+      },
+      calculateEphemeralMacKeyJwk: async (input) => {
+        const { privateKey, publicKey, sessionTranscriptBytes } = input
+        const ikm = p256
+          .getSharedSecret(TypedArrayEncoder.toHex(privateKey), TypedArrayEncoder.toHex(publicKey), true)
+          .slice(1)
+        const salt = Hasher.hash(sessionTranscriptBytes, 'sha-256')
+        const info = Buffer.from('EMacKey', 'utf-8')
+        const hk1 = hkdf(sha256, ikm, salt, info, 32)
+
+        return {
+          key_ops: ['sign', 'verify'],
+          ext: true,
+          kty: 'oct',
+          k: TypedArrayEncoder.toBase64URL(hk1),
+          alg: 'HS256',
+        }
+      },
+    },
+
+    cose: {
+      mac0: {
+        sign: async (input) => {
+          const { jwk, mac0 } = input
+          const { data } = mac0.getRawSigningData()
+          return await agentContext.wallet.sign({
+            data: Buffer.from(data),
+            key: getJwkFromJson(jwk as JwkJson).key,
+          })
+        },
+        verify: async (input) => {
+          const { mac0, jwk, options } = input
+          const { data, signature, alg } = mac0.getRawVerificationData(options)
+          return await verifyWithJwk({ jwk, signature, data, alg })
```

Comment:
does this still use the credo wallet to verify the signature? I can't see the context or something passed around. It is important. 

Aslo it's prob ok for now, but we should try to stay away from doing crypto operations outside of the wallet. Also with e.g. the p256 shared secret calcuation etc.. The more we can do in the wallet, the better.

So if you can provide a list of all cryptographic operations that are needed, we can look to move them to the wallet over time

Reply:
We just need the key derivation.